File: wearapp/src/main/java/com/thewizrd/simpleweather/fragments/SwipeDismissPreferenceFragment.java
package com.thewizrd.simpleweather.fragments;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.wear.widget.SwipeDismissFrameLayout;

import com.thewizrd.simpleweather.databinding.ActivitySettingsBinding;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public abstract class SwipeDismissPreferenceFragment extends PreferenceFragment {

    @IntDef({
            Toast.LENGTH_SHORT,
            Toast.LENGTH_LONG
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ToastDuration {
    }

    private Activity mActivity;
    private ActivitySettingsBinding binding;
    private SwipeDismissFrameLayout.Callback swipeCallback;

    public final Activity getParentActivity() {
        return mActivity;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mActivity = (Activity) context;
    }

    @Override
    public void onDetach() {
        mActivity = null;
        super.onDetach();
    }

    @Override
    public void onDestroy() {
        mActivity = null;
        super.onDestroy();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        onCreatePreferences(savedInstanceState);
    }

    @SuppressLint("RestrictedApi")
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        binding = ActivitySettingsBinding.inflate(inflater, container, false);
        View inflatedView = super.onCreateView(inflater, container, savedInstanceState);

        binding.swipeLayout.addView(inflatedView);
        binding.swipeLayout.setSwipeable(true);
        swipeCallback = new SwipeDismissFrameLayout.Callback() {
            @Override
            public void onDismissed(SwipeDismissFrameLayout layout) {
                if (mActivity != null) mActivity.onBackPressed();
            }
        };
        binding.swipeLayout.addCallback(swipeCallback);
        binding.swipeLayout.requestFocus();

        return binding.swipeLayout;
    }

    @Override
    public void onDestroyView() {
        binding.swipeLayout.removeCallback(swipeCallback);
        super.onDestroyView();
        binding = null;
    }

    public abstract void onCreatePreferences(@Nullable Bundle savedInstanceState);

    public void showToast(@StringRes int resId, @ToastDuration int duration) {
        if (mActivity != null) {
            mActivity.runOnUiThread(() -> {
                if (mActivity != null && isVisible()) {
                    Toast.makeText(mActivity, resId, duration).show();
                }
            });
        }
    }

    public void showToast(CharSequence message, @ToastDuration int duration) {
        if (mActivity != null) {
            mActivity.runOnUiThread(() -> {
                if (mActivity != null && isVisible()) {
                    Toast.makeText(mActivity, message, duration).show();
                }
            });
        }
    }
}

File: wearapp/src/main/java/com/thewizrd/simpleweather/fragments/SwipeDismissPreferenceFragment.kt
package com.thewizrd.simpleweather.fragments

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.lifecycle.lifecycleScope
import androidx.wear.widget.SwipeDismissFrameLayout
import com.thewizrd.shared_resources.utils.SettingsManager
import com.thewizrd.simpleweather.App
import com.thewizrd.simpleweather.databinding.ActivitySettingsBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext
import kotlin.coroutines.EmptyCoroutineContext

abstract class SwipeDismissPreferenceFragment : LifecyclePreferenceFragment() {
    @IntDef(Toast.LENGTH_SHORT, Toast.LENGTH_LONG)
    @Retention(AnnotationRetention.SOURCE)
    annotation class ToastDuration

    var parentActivity: Activity? = null
        private set
    protected val settingsManager: SettingsManager = App.instance.settingsManager

    private lateinit var binding: ActivitySettingsBinding
    private var swipeCallback: SwipeDismissFrameLayout.Callback? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        parentActivity = context as Activity
    }

    override fun onDetach() {
        parentActivity = null
        super.onDetach()
    }

    override fun onDestroy() {
        parentActivity = null
        super.onDestroy()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        onCreatePreferences(savedInstanceState)
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?
    ): View? {
        binding = ActivitySettingsBinding.inflate(inflater, container, false)
        val inflatedView = super.onCreateView(inflater, container, savedInstanceState)

        binding.swipeLayout.addView(inflatedView)
        binding.swipeLayout.isSwipeable = true
        swipeCallback = object : SwipeDismissFrameLayout.Callback() {
            override fun onDismissed(layout: SwipeDismissFrameLayout) {
                parentActivity?.onBackPressed()
            }
        }
        binding.swipeLayout.addCallback(swipeCallback)
        binding.swipeLayout.requestFocus()

        return binding.swipeLayout
    }

    override fun onDestroyView() {
        binding.swipeLayout.removeCallback(swipeCallback)
        super.onDestroyView()
    }

    abstract fun onCreatePreferences(savedInstanceState: Bundle?)

    fun showToast(@StringRes resId: Int, @ToastDuration duration: Int) {
        lifecycleScope.launch {
            if (parentActivity != null && isVisible) {
                Toast.makeText(parentActivity, resId, duration).show()
            }
        }
    }

    fun showToast(message: CharSequence?, @ToastDuration duration: Int) {
        lifecycleScope.launch {
            if (parentActivity != null && isVisible) {
                Toast.makeText(parentActivity, message, duration).show()
            }
        }
    }

    /**
     * Launches and runs the given runnable if the fragment is at least initialized
     * The action will be signalled to cancel if the fragment goes into the destroyed state
     * Note: This will run on the UI thread
     *
     * @param context additional to [CoroutineScope.coroutineContext] context of the coroutine.
     * @param block the coroutine code which will be invoked in the context of the viewLifeCycleOwner lifecycle scope.
     */
    fun runWithView(context: CoroutineContext = EmptyCoroutineContext,
                    block: suspend CoroutineScope.() -> Unit
    ) {
        runCatching {
            viewLifecycleOwner.lifecycleScope
        }.onFailure {
            // no-op
        }.onSuccess {
            it.launch(context = context, block = block)
        }
    }
}